streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java

```diff
@@ -111,15 +118,34 @@ public ProcessorNode build(String applicationId) {
     }
 
     private static class SourceNodeFactory extends NodeFactory {
-        public final String[] topics;
+        private final String[] topics;
+        public final Pattern pattern;
         private Deserializer keyDeserializer;
         private Deserializer valDeserializer;
+        private final SubscriptionUpdates subscriptionUpdates;
 
-        private SourceNodeFactory(String name, String[] topics, Deserializer keyDeserializer, Deserializer valDeserializer) {
+        private SourceNodeFactory(String name, String[] topics, Pattern pattern, Deserializer keyDeserializer, Deserializer valDeserializer, SubscriptionUpdates subscriptionUpdates) {
             super(name);
-            this.topics = topics.clone();
+            this.topics = topics != null ? topics.clone() : null;
+            this.pattern = pattern;
             this.keyDeserializer = keyDeserializer;
             this.valDeserializer = valDeserializer;
+            this.subscriptionUpdates = subscriptionUpdates;
```

Comment:
Seems that we always pass the same `subscriptionUpdates` object to all the `SourceNodeFactory`, could we instead overload `getTopics` with `getTopics(subscriptionUpdates.getUpdates())`, and trigger either one of the functions based on whether it contains the pattern or not?